### DISCUSSION TITLE
astro:actions does not support cancelling / aborting / timeout (AbortController)

### DISCUSSION THREAD

**Discussion by smnbbrv** [withastro/roadmap#1047] [Proposal]
### Body

# Summary

It is currently impossible to cancel the call to actions.

# Background & Motivation

In some scenarios e.g. typeahead search it is important to cancel already irrelevant calls to the backend. Another scenario would be to cancel a long running task.

# Goals

A way to abort the call with AbortController or another way to do it via actions API. Would be also perfect if one can specify additionally a timeout (however having AbortController already gives a way to implement it manually)

# Example

Client:

```ts
import { actions } from 'astro:actions';

const { signal, abort } = new AbortController();

const result = await actions.search({ query: '123' }, { abort: signal })

setTimeout(abort, 1000)
```

Server:

```ts
    handler: async ({ query }, { aborted }) => {
       // aborted would be e.g. a Promise that is resolved when the request is cancelled
       aborted.then()
```

**Comment by florian-lefebvre** [comment]
A ditry workaround on the client is to patch `fetch`:

```diff
import { actions } from 'astro:actions';

const { signal, abort } = new AbortController();

+ const originalFetch = window.fetch;
+ window.fetch = (input, init) => {
+    init ??= {};
+    init.signal = signal;
+    return originalFetch(input, init);
+};

const result = await actions.search({ query: '123' })

+window.fetch = originalFetch;

setTimeout(abort, 1000)
```

**Comment by florian-lefebvre** [comment]
FYI I'm working on https://github.com/withastro/astro/pull/12721 which should unlock more advanced usages

**Reply by smnbbrv** [reply]
@florian-lefebvre is there a working example for how to achieve the desired functionality? Should it be provided in the docs?

**Reply by florian-lefebvre** [reply]
There's an example at https://docs.astro.build/en/reference/modules/astro-actions/#getactionpath

**Reply by smnbbrv** [reply]
@florian-lefebvre thanks! This makes it possible, however this requires to refuse the official api of the actions in favour of native fetch in order to simply cancel the request.

**Reply by florian-lefebvre** [reply]
I just reopened the conversation :+1:
